chore: remove gulp-git dependency

marionebl opened the pull request:
*  gets rid of gulp dependency
*  gets rid of gulp-git dependency
*  solves gulp-related deprecation warnings
*  improves install times considerably
*  use child_process.exec for git commands

---
- [x] fix coverage
- [x] fix windows support
- [x] restore cwd switching to `commit` and `log`

marionebl commented:
Hey there, this should be done now. Tell me what you think – I'd squash this if your fine with the idea

LinusU commented on src/cli/strategies/git-cz.js:
```diff
@@ -47,20 +47,32 @@ function gitCz(rawGitArgs, environment, adapterConfig) {
-      });
-      
+      let cleanError = new Error('Error: No files added to staging! Did you forget to run git add?');
+      throw cleanError;
```
Why assign the error to a variable first?

marionebl replied:
Fixed

LinusU commented on src/commitizen/adapter.js:
```diff
@@ -78,6 +78,7 @@ function getNearestNodeModulesDirectory(options) {
   let nodeModulesDirectories = findNodeModules(options);
   
    // Make sure we find a node_modules folder
+  /* istanbul ignore else */
```
Should we really add the ignore things? Potentially we could write tests for it, isn't this just tricking us into thinking we have a high code coverage?

I usually only add it when there should be 100% impossible to reach it. example:

```js
function test (a) {
  let n
  switch (a) {
    case 'one': n = 1; break
    case 'two': n = 2; break
    default: throw new Error('...')
  }

  switch (a) {
    case 'one': n += 10; break
    case 'two': n += 20; break
    /* istanbul ignore next */
    default: throw new Error('...')
  }

  return n
}
```

marionebl replied:
I thought ignoring the error cases was okay, testing them could prove challenging with the unit tests running on actual git commands instead of a mock layer.

LinusU replied:
I agree that we shouldn't add the tests now, but isn't it better to not ignore them now, so that we will be reminded to do it at some point? and to not falsely advertise 100% test coverage?

marionebl commented:
@LinusU I removed all the ignore comments and refactored [git/commit](https://github.com/commitizen/cz-cli/pull/256/commits/6f5badac0e681e97a35a3ef17532ec84d409e041) to a way more concise format, causing the tests to fulfill the coverage thresholds.

LinusU commented on src/cli/strategies/git-cz.js:
```diff
@@ -47,20 +47,31 @@ function gitCz(rawGitArgs, environment, adapterConfig) {
-        // console.log('commit happened');
-      });
-      
+      throw new Error('Error: No files added to staging! Did you forget to run git add?');
```
You can remove the `Error: ` part from the start of the message, the error class will add that when casting to a string.

LinusU commented:
Can you rebase this on latest master now that your two other pull requests got merged? Remember to remove those commit when rebasing so that they don't get added twice...

marionebl commented:
Rebased master and fixed up all commits, reworded base commit from "chore" to "refactor" type.

marionebl commented:
Ping!

marionebl commented:
Any news on this?

LinusU commented:
Sorry, I've been away celebrating Swedish midsummer 🇸🇪 🌳 🍸 

I'll try and look at this shortly

LinusU commented:
It doesn't seem like you are removing gulp and gulp-git from the package.json, could you do that?

LinusU commented:
Instead of using child_process.exec, would it be better to use `execFile` or `spawn` so that we can give the arguments as an array instead as a string. That way we don't have to worry about quoting at all which I think would improve that part of the code base.

marionebl commented:
Removed gulp and gulp-git from `.dependencies`. I'd stay with `child_process`'s `exec` for now because `git/commit.js` needs

*  `normalizeCommitMessage` for quote normalization anyway
*   would need to parse `options.args` somehow, which currently allows injection of arbitrary args
*   takes on some complexity for error handling when changing to `spawn`

LinusU commented on src/commitizen/staging.js:
```diff
@@ -1,20 +1,19 @@
-      stagingIsClean = false;
-    } else {
-      stagingIsClean = true;
+  exec('git diff --cached --name-only', {
```
Should we keep `--no-pager` here?

marionebl replied:
I think `--no-pager` isn't needed here, as the succeeding tests at [test/tests/staging](https://github.com/commitizen/cz-cli/blob/07c10335c22143f6421675a20e85a13e424aae23/test/tests/staging.js#L60) show. I guess the option was intended to guard against counting the length of the pager `(END)` string. According to my tests `git diff` output is not paged anyway in non-tty processes.

LinusU commented:
> I'd stay with child_process's exec for now because...

Fair enough 👌 

One last question but otherwise I'm okay on landing this.

@jimthedev any feelings on this?

jimthedev commented:
I've been listening and at this point I'm ok with landing this. You've
covered all of the major topics. One thing we might want to do is check the
old vs new output to stdout to make sure that any changes in that regard
are documented. It is tricky to tell how other projects might be using that
output, but if it is a significant difference then we may want to release
it as a major version update.

Otherwise I'm totally on board with removing these.

On June 29, 2016 at 12:52:42 AM, Linus Unnebäck (notifications@github.com)
wrote:

> I'd stay with child_process's exec for now because...
>
> Fair enough 👌
>
> One last question but otherwise I'm okay on landing this.
>
> @jimthedev <https://github.com/jimthedev> any feelings on this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commitizen/cz-cli/pull/256#issuecomment-229263137>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAGpitqXEaUe1cKYMAJhqzEn-5qFPoIcks5qQggqgaJpZM4Izqgw>
> .
>

marionebl commented:
Concerning `stdout`, `stderr` and exit codes:

**Changes**
* [cli/strategies/git-cz.js](https://github.com/commitizen/cz-cli/pull/256/files#diff-23495cac1ee02ad0f5c48671e7498c16R56) Staging check now exits with code 1 if the stage is clean. `stderr` output now contains a stack trace below the error message. 
* [cli/strategies/git-cz.js](https://github.com/commitizen/cz-cli/pull/256/files#diff-23495cac1ee02ad0f5c48671e7498c16R71) Committing now exits with code 1 if committing fails instead of failing silently. `stderr` output now contains a stack trace.
* [cli/strategies/git.js](https://github.com/marionebl/cz-cli/blob/dbf16d8355f2ca93dedfa8385b437b189ec63a37/src/cli/strategies/git.js#L19) Committing now exits with code 1 if committing fails instead of just printing an error message. `stderr` output now contains a stack trace below the error message. 

**Important unchanged stuff**
* [git/commit.js](https://github.com/commitizen/cz-cli/pull/256/files#diff-d62e50e243d79d33fb86519ae515265bR31) still respects the `quiet` option. Output should be exactly the same as with [gulp-git](https://github.com/stevelacy/gulp-git/blob/master/lib/commit.js#L68) in success scenarios. When committing errors the `git-cz` strategy kicks in and the changes there apply.

marionebl commented:
Major version bump for these changes?

marionebl commented:
Any news?

LinusU commented:
Sorry again for delays. I would prefer not to make a major bump since I don't see that this would have an impact on current users...

I have a really hard time imagining that changing the error output would be a problem, unless someone is doing exact string comparison on errors which seems unlikely.

Someone could potentially be relaying on the exit code always being `0`. However, I don't think that many users, if any at all, does this since it's very very non-standard to always exit with a success code. If I where do write a script that expected any failures I would first of all assume that this utility actually exited with an error code, and I would have opened an issue when I found out that it didn't.

I'm okay with merging this now and releasing as a patch release. @jimthedev should have final say though, if you just give the go ahead I will happily merge and cut a new release 👌

marionebl commented:
Hey @LinusU @jimthedev, can we go forward on this?

jimthedev commented:
I am on board with @LinusU's plan.

LinusU commented:
Great, thank you for your work @marionebl!
